### PR TITLE
MattT/APPEALS-8467: Only Show Appeals in Completed Tab That Have Not Been Returned to CSP

### PR DESCRIPTION
**Pull request by ThorntonMatthew**
Resolves [APPEALS-8467](https://vajira.max.gov/browse/APPEALS-8467)

### Description
It was possible that if an appeal was returned to the VHA CSP from BVA Intake, a single appeal could have an entry (or entries) in the VHA CSP's Completed Tab in addition to one in its Unassigned/In Progress tab. This PR aims to limit an appeal to only being able to be represented a single time in the VHA CSP's queue.

### Acceptance Criteria
- [ ] The Completed tab only shows appeals that have been completed and not returned to CSP (unless the newest CSP task has been completed).

### Testing Plan
1. Log into Caseflow as an Intake user (ex: `BVAISHAW`). Intake an appeal with an issue with the following attributes:

   - [ ] Benefit Type: Veterans Health Administration
   - [ ] Issue Category: Caregiver | ...
   
Take note of the veteran file number you've used.
   
2. Switch to a VHA CSP user (ex: `CAREGIVERUSER`) and navigate to the VHA CSP's queue:
`/organizations/vha-csp`

The appeal should be in the "Unassigned" tab at this time.

3. Fire up the Rails console, and simulate the VHA CSP sending the appeal back to BVA Intake:
```ruby
# Appeal
a = Appeal.find_by_uuid("<Appeal-UUID>")

# PreDocketTask
pdt = a.tasks.find_by(type: PreDocketTask.name)

# VhaDocumentSearchTask
vdst = a.tasks.find_by(type: VhaDocumentSearchTask.name)

# Complete the VhaDocumentSearchTask, sending the appeal back to BVA Intake
vdst.completed!
```

Task tree at this time:

```ruby
a.treee
=> 
Appeal 1637 (D 220731-1637 Original)  │ ID   │ STATUS    │ ASGN_BY  │ ASGN_TO             │ UPDATED_AT              │
└── RootTask                          │ 7111 │ on_hold   │          │ Bva                 │ 2022-08-18 19:15:07 UTC │
    └── PreDocketTask                 │ 7112 │ assigned  │          │ BvaIntake           │ 2022-08-23 01:28:51 UTC │
        ├── VhaDocumentSearchTask     │ 7113 │ completed │ BVAISHAW │ VhaCaregiverSupport │ 2022-08-22 17:34:10 UTC │                                      
```

4. Navigate to the "Completed" tab in VHA CSP's queue and make sure this appeal is now there.

5. Simulate BVA Intake sending the appeal back to VHA CSP in the Rails console:

```ruby
# This command relies on the ones from step 3
new_vdst = VhaDocumentSearchTask.create!(appeal: a, parent: pdt, assigned_to: VhaCaregiverSupport.singleton)
```

Task tree:

```ruby
a.treee
=>
Appeal 1637 (D 220731-1637 Original)  │ ID   │ STATUS    │ ASGN_BY  │ ASGN_TO             │ UPDATED_AT              │
└── RootTask                          │ 7111 │ on_hold   │          │ Bva                 │ 2022-08-18 19:15:07 UTC │
    └── PreDocketTask                 │ 7112 │ on_hold   │          │ BvaIntake           │ 2022-08-23 01:28:51 UTC │
        ├── VhaDocumentSearchTask     │ 7113 │ completed │ BVAISHAW │ VhaCaregiverSupport │ 2022-08-22 17:34:10 UTC │
        └── VhaDocumentSearchTask     │ 7121 │ assigned  │          │ VhaCaregiverSupport │ 2022-08-23 01:28:41 UTC │
```

6. The appeal should now be in the "Unassigned" tab and **not** in the "Completed" tab.

7. Set the newest VhaDocumentSearchTask to `in_progress`.

```ruby
new_vdst.in_progress!

a.treee
=>
Appeal 1637 (D 220731-1637 Original)  │ ID   │ STATUS       │ ASGN_BY  │ ASGN_TO             │ UPDATED_AT              │
└── RootTask                          │ 7111 │ on_hold      │          │ Bva                 │ 2022-08-18 19:15:07 UTC │
    └── PreDocketTask                 │ 7112 │ on_hold      │          │ BvaIntake           │ 2022-08-23 01:28:51 UTC │
        ├── VhaDocumentSearchTask     │ 7113 │ completed    │ BVAISHAW │ VhaCaregiverSupport │ 2022-08-22 17:34:10 UTC │
        └── VhaDocumentSearchTask     │ 7121 │ in_progress  │          │ VhaCaregiverSupport │ 2022-08-23 01:28:41 UTC │
```

8. The appeal should now be in the "In Progress" tab and **not** in the "Unassigned" or "Completed" tabs.

9. Go crazy!

```ruby
# Don't forget to close out this guy first!
new_vdst.completed!

(1..20).each do
   new_vdst = VhaDocumentSearchTask.create!(appeal: a, parent: pdt, assigned_to: VhaCaregiverSupport.singleton)
   new_vdst.completed!
end
```

Task tree:

```ruby
                                      ┌─────────────────────────────────────────────────────────────────────────────┐
Appeal 1637 (D 220731-1637 Original)  │ ID   │ STATUS    │ ASGN_BY  │ ASGN_TO             │ UPDATED_AT              │
└── RootTask                          │ 7111 │ on_hold   │          │ Bva                 │ 2022-08-18 19:15:07 UTC │
    └── PreDocketTask                 │ 7112 │ assigned  │          │ BvaIntake           │ 2022-08-23 01:51:33 UTC │
        ├── VhaDocumentSearchTask     │ 7113 │ completed │ BVAISHAW │ VhaCaregiverSupport │ 2022-08-22 17:34:10 UTC │
        ├── VhaDocumentSearchTask     │ 7121 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:50:40 UTC │
        ├── VhaDocumentSearchTask     │ 7122 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:30 UTC │
        ├── VhaDocumentSearchTask     │ 7123 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:30 UTC │
        ├── VhaDocumentSearchTask     │ 7124 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:31 UTC │
        ├── VhaDocumentSearchTask     │ 7125 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:31 UTC │
        ├── VhaDocumentSearchTask     │ 7126 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:31 UTC │
        ├── VhaDocumentSearchTask     │ 7127 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:31 UTC │
        ├── VhaDocumentSearchTask     │ 7128 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:31 UTC │
        ├── VhaDocumentSearchTask     │ 7129 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:31 UTC │
        ├── VhaDocumentSearchTask     │ 7130 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:31 UTC │
        ├── VhaDocumentSearchTask     │ 7131 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:31 UTC │
        ├── VhaDocumentSearchTask     │ 7132 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:32 UTC │
        ├── VhaDocumentSearchTask     │ 7133 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:32 UTC │
        ├── VhaDocumentSearchTask     │ 7134 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:32 UTC │
        ├── VhaDocumentSearchTask     │ 7135 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:32 UTC │
        ├── VhaDocumentSearchTask     │ 7136 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:32 UTC │
        ├── VhaDocumentSearchTask     │ 7137 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:32 UTC │
        ├── VhaDocumentSearchTask     │ 7138 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:32 UTC │
        ├── VhaDocumentSearchTask     │ 7139 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:32 UTC │
        ├── VhaDocumentSearchTask     │ 7140 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:33 UTC │
        └── VhaDocumentSearchTask     │ 7141 │ completed │          │ VhaCaregiverSupport │ 2022-08-23 01:51:33 UTC │
                                      └─────────────────────────────────────────────────────────────────────────────┘
```

Even with this ridiculous task tree, you should only see this appeal listed once in the entirety of VHA CSP's queue, and it should be on the "Completed" tab.

10. Ensure that the "Completed" tab does not include this appeal any longer once the newest CSP task's `closed_at` value is 7+ days in the past:

```ruby
new_vdst.update(closed_at: (7.days + 1.minute).before)
```

It should no longer appear on any of the tabs.

Bonus:

Intake/create multiple appeals and place them into various states, and see that they appear on the expected tabs.

- [ ] For higher-risk changes: [Deploy the custom branch to UAT to test](https://github.com/department-of-veterans-affairs/appeals-deployment/wiki/Applications---Deploy-Custom-Branch-to-UAT)

### CODE REVIEW

File: spec/feature/queue/caregiver_support_queue.rb [hunk 1]
```diff
@@ -16,15 +16,15 @@
     let!(:num_in_progress_rows) { 9 }
     let!(:num_completed_rows) { 5 }
 
-    let!(:vha_caregiver_unassigned_tasks) { 
+    let!(:vha_caregiver_unassigned_tasks) do
       create_list(:vha_document_search_task, num_unassigned_rows, :assigned, assigned_to: csp_org)
-    }
```
Comment (ThorntonMatthew) [inline spec/feature/queue/caregiver_support_queue.rb:21]: All of the changes in this file are from rubocop fixing issues that came in from the feature branch.

File: app/models/queue_tabs/vha_caregiver_support_completed_tasks_tab.rb [hunk 2]
```diff
@@ -16,7 +16,16 @@ def description
...
+
+    # For every appeal that has a recently completed VHA CSP task, determine
+    # which VHA CSP task is actually the most recent for the entire appeal
+    # so that the appeal doesn't show up multiple times in the VHA CSP's queue.
+    most_recent_csp_tasks_per_appeal = Task.where(parent_id: parent_task_ids, assigned_to: assignee)
+      .group(:appeal_id)
+      .maximum(:id)
+
+    Task.where(id: most_recent_csp_tasks_per_appeal.values).recently_completed
   end
```
Comment (ajspotts) [inline app/models/queue_tabs/vha_caregiver_support_completed_tasks_tab.rb:29]: @ThorntonMatthew wondering if it's worth extracting some of this (if not all) to [queue_tab](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/queue_tab.rb) for reusability as it seems there would most likely be other pre docket queues (CAMO for instance) where we'd want this same functionality for the Completed tab

File: app/models/queue_tab.rb [hunk 3]
```diff
@@ -119,6 +119,18 @@ def recently_completed_tasks
     Task.includes(*task_includes).visible_in_queue_table_view.where(assigned_to: assignee).recently_completed
   end
 
+  # Recently completed tasks that do not have younger sibling tasks
+  # (tasks with the same parent task but have been created more recently) with the same assignee.
+  def recently_completed_tasks_without_younger_siblings
```
Comment (ajspotts) [inline app/models/queue_tab.rb:124]: 🙌 i like it, thanks for refactoring that!

**Review by ajspotts — APPROVED** [review]
lgtm!

**Pull request merged**